sdk/communication/AzureCommunicationChat/Source/Signaling/CommunicationSignalingClient.swift

```diff
@@ -106,13 +111,64 @@ class CommunicationSignalingClient {
 }
 
 class CommunicationSkypeTokenProvider: NSObject, TrouterSkypetokenProvider {
-    var skypeToken: String?
-    func getSkypetoken(_: Bool) -> String! {
-        return skypeToken
-    }
+    /// The cached skypeToken.
+    var token: String
+
+    /// The CommunicationTokenCredential.
+    var credential: CommunicationTokenCredential
+
+    /// Called from getSkypetoken, handle token errors here.
+    var tokenRefreshHandler: TrouterTokenRefreshHandler
+
+    /// Current number of token fetch retries.
+    var tokenRetries: Int
 
-    init(skypeToken: String? = nil) {
-        self.skypeToken = skypeToken
+    /// Max number of token retries allowed.
+    let MAX_TOKEN_RETRIES: Int = 3
+
+    /// Return the cached token, will attempt to refresh the token if forceRefresh is true.
+    func getSkypetoken(_ forceRefresh: Bool) -> String! {
+        if forceRefresh {
+            tokenRetries += 1
+
+            // We have to return the token but don't attempt to refresh again
+            // Pass true to the callback to signal that we should stop the connection
+            if tokenRetries > MAX_TOKEN_RETRIES {
+                tokenRefreshHandler(true)
+                return token
+            }
+
+            // Fetch new token
+            credential.token { token, error in
+                // Let callback know we are attempting to refresh
+                self.tokenRefreshHandler(false)
+                guard let newToken = token?.token else {
+                    return
+                }
```

Comment:
Why are you not doing anything with the error here? If token is nil, presumably there would be an error.

Reply:
Updated, to pass the error to chatClient